### PR TITLE
Change mastodon lib to Megalodon , add misskey and calckey support

### PR DESCRIPTION
**Pull request by aneillans**
This PR is a large one I'm afraid.  Would recommend squashing this on merge!

- NodeJS bumped up
- A LOT of dependencies upgraded, including TypeScript as it was seriously of date
- Swapped out the lib for Mastodon to be Megalodon and added support for MissKey and Calckey (in theory Megalodon would support Pleroma too but I don't have a test account on anything for that to build against yet) 

The auth for MissKey etc is all done local, so no server side bits at all (did look at using the auth server but changed my mind as it's not really needed tbh for OAuth and just adds an additional maintenance overhead for no benefit as not offloading posting schedules and such).

### CODE REVIEW

**Comment by mvdicarlo** [comment]
I'm going to need more time to PR this more thoroughly and test it manually.;

**Comment by aneillans** [comment]
> I'm going to need more time to PR this more thoroughly and test it manually.;

Yeah its a big one! If I get time, will keep it in sync with develop. I have a tester who has flagged an issue with CalcKey support which I need to bug fix too.

**Comment by aneillans** [comment]
I've found an underlying bug in Megalodon with Calckey, and and am retesting the core functionality with just Misskey - will update shortly.

There's a linked PR now for Bluesky (needs the same Dependency changes as this lot so was a lot easier to just base off this branch).

Despite the linked dependency on the other PR, I'd still be tempted to say these were worth merging down (will update for .23 and rebase both) but mark them as Alpha and I'll fix up as needed.

**Comment by mvdicarlo** [comment]
Forgot to officially post the comments I had. Still getting used to github prs I guess.

File: electron-app/tsconfig.json [hunk 1]
```diff
@@ -5,6 +5,7 @@
     "removeComments": true,
     "emitDecoratorMetadata": true,
     "experimentalDecorators": true,
+    "esModuleInterop": true,
```
[the remark this answers is not in the document]
Comment (aneillans) [reply electron-app/tsconfig.json:8]: Change is required for compatibility with Megalodon but get what you mean. Will keep in mind for future.

File: ui/src/views/submissions/submission-template-select/SubmissionTemplateSelect.tsx [hunk 2]
```diff
@@ -107,7 +108,8 @@ export default class SubmissionTemplateSelect extends React.Component<Props> {
   handleSelect = (value: string | number | LabeledValue, option: React.ReactElement<any>) => {
     const [type, id] = value.toString().split(':');
     
-    let parts: Record<string, SubmissionPart<any>> = {};
+//    let parts: Record<string, SubmissionPart<any>> = {};
```
[the remark this answers is not in the document]
Comment (aneillans) [reply ui/src/views/submissions/submission-template-select/SubmissionTemplateSelect.tsx:111]: Yup. I should really work out the correct typing, but ... for now this works.

File: ui/package.json [hunk 3]
```diff
@@ -14,7 +14,9 @@
     "antd": "^3.26.20",
     "axios": "^0.21.4",
     "file-saver": "^2.0.5",
+    "helpers": "^0.0.6",
```
[the remark this answers is not in the document]
Comment (aneillans) [reply ui/package.json]: Good catch; doesn't look like it's actually needed... removed.

File: ui/src/websites/misskey/MissKey.tsx [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import React from 'react';
+import { WebsiteSectionProps } from '../form-sections/website-form-section.interface';
+import GenericFileSubmissionSection from '../generic/GenericFileSubmissionSection';
+import { GenericSelectProps } from '../generic/GenericSelectProps';
+import GenericSubmissionSection from '../generic/GenericSubmissionSection';
+import { LoginDialogProps } from '../interfaces/website.interface';
+import { WebsiteImpl } from '../website.base';
+import MissKeyLogin from './MissKeyLogin';
+
+export class MissKey extends WebsiteImpl {
```
[the remark this answers is not in the document]
Comment (mvdicarlo) [reply ui/src/websites/misskey/MissKey.tsx:18]: So should I just hold off? So you have the time to refactor and bring all the work you've done together with cleanup?

File: electron-app/src/server/websites/misskey/misskey.service.ts [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+      }      
+     }
+
+      // Update the post content with the Tags if any are specified - for MissKey, we need to append 
+      // these onto the post, *IF* there is character count available.
+      if (data.tags.length > 0) {
+        status += "\n\n";
+      }
+
+      data.tags.forEach(tag => {
```
[the remark this answers is not in the document]
Comment (aneillans) [reply electron-app/src/server/websites/misskey/misskey.service.ts]: Is that commit correct?

File: ui/src/websites/misskey/MissKey.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (aneillans) [reply ui/src/websites/misskey/MissKey.tsx:18]: Not sure; I'd be tempted to say no, given that is going to be dependant on when the other package merge down too - at the minute there's a bug around image handling for CalcKey / MissKey (I'm tempted to disable file handling in PostyBirb for that for now) which would allow this to move - then we can update after once they have accepted. Or worst case vendor the lib.

**Comment by aneillans** [comment]
Rebased on the current .24 and tested.

File: electron-app/src/server/websites/mastodon/mastodon.service.ts [hunk 6]
```diff
@@ -276,64 +288,59 @@ export class Mastodon extends Website {
...
       return Promise.reject(
-        this.createPostResponse({ message: post.data.error, additionalInfo: post.data }),
+        this.createPostResponse({ message: err.message }),
       );
-    }
-
-    return this.createPostResponse({ source: post.data.url });
-  }
-
-  formatTags(tags: string[]) {
```
Comment (mvdicarlo) [inline electron-app/src/server/websites/mastodon/mastodon.service.ts:325]: Bring this codeblock back in to simplify tag parsing.

File: electron-app/src/server/websites/misskey/misskey.service.ts [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+        maxLength: 100,
+      });
+
+      // Update the post content with the Tags if any are specified - for MissKey, we need to append 
+      // these onto the post, *IF* there is character count available.
+      if (tags.length > 0) {
+        status += "\n\n";
+      }
+
+      tags.forEach(tag => {
```
Comment (mvdicarlo) [inline electron-app/src/server/websites/misskey/misskey.service.ts:215]: Same comment here as in mastodon service.

File: electron-app/src/server/websites/misskey/misskey.service.ts [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+      maxLength: 100,
+    });
+
+    // Update the post content with the Tags if any are specified - for MissKey, we need to append 
+    // these onto the post, *IF* there is character count available.
+    if (tags.length > 0) {
+      status += "\n\n";
+    }
+
+    tags.forEach(tag => {
```
Comment (mvdicarlo) [inline electron-app/src/server/websites/misskey/misskey.service.ts:279]: Here too

File: electron-app/src/server/websites/pixelfed/pixelfed.service.ts [hunk 9]
```diff
@@ -180,85 +195,69 @@ export class Pixelfed extends Website {
...
       // these onto the post, *IF* there is character count available.
-      if (tags.length > 0) {
-        form.status += '\n\n';
+      if (data.tags.length > 0) {
+        status += "\n\n";
       }
 
-      tags.forEach((tag) => {
-        let remain = maxChars - form.status.length;
+      data.tags.forEach(tag => {
```
Comment (mvdicarlo) [inline electron-app/src/server/websites/pixelfed/pixelfed.service.ts:229]: And here.

I just don't want any tagging regressions. See formatTags in the code diff.

File: electron-app/src/server/websites/pixelfed/pixelfed.service.ts [hunk 10]
```diff
@@ -180,85 +195,69 @@ export class Pixelfed extends Website {
...
+        lastId = result.data.id;
+        let res = result.data as Entity.Status;
+        return this.createPostResponse({ source: res.url });
+      }).catch((err: Error) => {
         return Promise.reject(
-          this.createPostResponse({ message: post.data.error, additionalInfo: post.data }),
+          this.createPostResponse({ message: err.message }),
         );
-      }
+      })
```
Comment (mvdicarlo) [inline electron-app/src/server/websites/pixelfed/pixelfed.service.ts:253]: Needs to return or await.
Looks like this just falls through.

File: electron-app/src/server/auth.guard.ts [hunk 11]
```diff
@@ -6,7 +6,7 @@ export class AuthGuard implements CanActivate {
   canActivate(context: ExecutionContext): boolean | Promise<boolean> | Observable<boolean> {
     const request = context.switchToHttp().getRequest();
     return (
-      request.headers.authorization === global.AUTH_ID || request.query.auth === global.AUTH_ID
+      request.headers.authorization === global.AUTH_ID || request.query.auth === global.AUTH_ID || request.path === "/misskey/display"
```
Comment (mvdicarlo) [inline electron-app/src/server/auth.guard.ts:9]: I don't see why this change is necessary.
It looks like you pass in the auth key in the login panel.

File: electron-app/src/server/websites/misskey/misskey.service.ts [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+
+      M.postStatus(status, statusOptions).then((result) => {
+        lastId = result.data.id;
+        let res = result.data as Entity.Status;
+        return this.createPostResponse({ source: res.url });
+      }).catch((err: Error) => {
+        return Promise.reject(
+          this.createPostResponse({ message: err.message }),
+        );
+      })
```
Comment (mvdicarlo) [inline electron-app/src/server/websites/misskey/misskey.service.ts:239]: Needs a return or I don't think this will actually return as you expect.

File: electron-app/src/server/websites/mastodon/mastodon.service.ts [hunk 13]
```diff
@@ -276,64 +288,59 @@ export class Mastodon extends Website {
...
-        .map((tag) => tag.replace(/[^a-z0-9]/gi, ' '))
-        .map((tag) =>
-          tag
-            .split(' ')
-            // .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
-            .join(''),
-        ),
-      { spaceReplacer: '' },
-    ).map((tag) => `#${tag}`);
+    })
```
Comment (mvdicarlo) [inline electron-app/src/server/websites/mastodon/mastodon.service.ts:342]: Needs a return or await.
This would just immediately fall through to line 343.

File: electron-app/src/server/websites/mastodon/mastodon.service.ts [hunk 14]
```diff
@@ -204,66 +200,82 @@ export class Mastodon extends Website {
...
+        return this.createPostResponse({ source: res.url });
+      }).catch((err: Error) => {
         return Promise.reject(
-          this.createPostResponse({ message: post.data.error, additionalInfo: post.data }),
+          this.createPostResponse({ message: err.message }),
         );
-      } else {
-        return this.createPostResponse({ source: post.data.url });
-      }
+      })
```
Comment (mvdicarlo) [inline electron-app/src/server/websites/mastodon/mastodon.service.ts:278]: Needs a return

**Review by mvdicarlo — COMMENTED** [review]
Hopefully the last of the reviews.
There are some build warnings in the ui about unused vars, but not critically blocking.

Some questions and code issues I saw. But should be the last of it. Thanks for your patience.

File: electron-app/src/server/auth.guard.ts (hunk 11 above)
Comment (aneillans) [reply electron-app/src/server/auth.guard.ts:9]: MissKey has a callback where the code is presented on a locally hosted page within PostyBirb, so we need to add this to allow the page to be accessed. Not automagically handled to avoid creating an X-site attack vector.

File: electron-app/src/server/websites/pixelfed/pixelfed.service.ts (hunk 10 above)
Comment (aneillans) [reply electron-app/src/server/websites/pixelfed/pixelfed.service.ts]: Resolved

File: electron-app/src/server/websites/misskey/misskey.service.ts (hunk 12 above)
Comment (aneillans) [reply electron-app/src/server/websites/misskey/misskey.service.ts:239]: Resolved

File: electron-app/src/server/websites/mastodon/mastodon.service.ts (hunk 13 above)
Comment (aneillans) [reply electron-app/src/server/websites/mastodon/mastodon.service.ts:342]: Resolved

File: electron-app/src/server/websites/mastodon/mastodon.service.ts (hunk 14 above)
Comment (aneillans) [reply electron-app/src/server/websites/mastodon/mastodon.service.ts]: Resolved

File: electron-app/src/server/websites/pixelfed/pixelfed.service.ts (hunk 9 above)
Comment (aneillans) [reply electron-app/src/server/websites/pixelfed/pixelfed.service.ts]: Resolved

File: electron-app/src/server/websites/misskey/misskey.service.ts (hunk 8 above)
Comment (aneillans) [reply electron-app/src/server/websites/misskey/misskey.service.ts:268]: Resolved

File: electron-app/src/server/websites/misskey/misskey.service.ts (hunk 7 above)
Comment (aneillans) [reply electron-app/src/server/websites/misskey/misskey.service.ts:211]: Resolved

File: electron-app/src/server/websites/mastodon/mastodon.service.ts (hunk 6 above)
Comment (aneillans) [reply electron-app/src/server/websites/mastodon/mastodon.service.ts:325]: Resolved

**Comment by aneillans** [comment]
Tested and working with latest MissKey version; does have an issue with file uploads on FireFish (calckey renamed to this) - which I'm looking at with Megalodon.

**Comment by leaftail1880** [comment]
Will be typescript version replaced to one for whole project?
Are you planning to replace tslint to eslint?

**Comment by aneillans** [comment]
> Will be typescript version replaced to one for whole project? Are you planning to replace tslint to eslint?
> 
> I has that in my plans so if not ill open new pr

Wouldn't look to do it under this pr, as it's already rather too large - but yeah, would be good to get it moved, given tslint is deprecated!

**Comment by leaftail1880** [comment]
> > Will be typescript version replaced to one for whole project? Are you planning to replace tslint to eslint?
> > I has that in my plans so if not ill open new pr
> 
> Wouldn't look to do it under this pr, as it's already rather too large - but yeah, would be good to get it moved, given tslint is deprecated!

I want to do that, but i dont sure what things will be changed in your pr so i cant start mine

**Comment by mvdicarlo** [comment]
Still missing some awaits, and I don't think you're returning like you are intending to. But to move this forward, I am going to to merge this to unblock you.

I can fix any lingering issues when I am back from vacation.

**Comment by mvdicarlo** [comment]
If you are ready for this to merge, reply here and I'll get it in.

**Comment by aneillans** [comment]
> If you are ready for this to merge, reply here and I'll get it in.

You were right, found a couple of missing awaits. Should be good to merge.
